#### **Introducción a la Estadística**

En este tutorial vamos a entender un poco de que se trata la estadística y a repasar algunos conceptos que serán útiles en temas que se verán más adelante. En resumen, estos son los temas que exploraremos en este tutorial:

1. ¿Qué es la Estadística?
2. Entendiendo el concepto de variación
3. Introducción a la estadística descriptiva
4. Medidas de Tendencia Central
5. Medidas de Dispersión

¡Ahora sí! ¡Empecemos!

Abróchate el cinturón y disfruta del viaje. Se esperan algunas turbulencias, pero haré lo posible para que sea un viaje seguro.

#### **¿Qué es la estadística?**

En términos muy simples. la estadística no hace más que estudiar la variación. Ajá, pero entonces, ¿qué es eso de la variación?

A medida que vayas viendo temas más avanzados de estadística vas a notar que se habla mucho de variación (por ej. el modelo que mejor explica la variación). Así que comencemos entendiendo este concepto. 

La variación forma parte de la vida. Las personas tienen opiniones diferentes, dos lapiceros de la misma marca y mismo color tienen tonos de tintas diferentes (auqnue tal vez imperceptibles para el ojo humano), la comida que cocinas (o te cocinan) no siempre tendrá la misma cantidad de sal (a veces tendrá un poco más, a veces tendrá un poco menos), dos computadoras del mismo modelo, misma marca, mismos specs no tendrá el mismo desempeño (aunque de nuevo, tal vez no lo puedas percibir). En fin, creo que vas entiendo por donde voy.

Lo anterior significa que **la estadística existe proque la variación existe**. Si de repente todos pensáramos igual no tendría sentido hacer encuestas de opinión (para qué, si todos dirán lo mismo). Si todos tuviésemos la misma capacidad, recursos y ambiente para aprender, no tendría sentido sacar un promedio del desempeño para un curso (¿Para qué, si al final todos sacarán lo mismo?) y asú sucesivamente.

Por tanto, la estadística lo que hace es **usar métodos y técnicas para poder estudiar esa variación y sacar conclusiones.** Por ejemplo, las encuestas de preferencia electoral se hacen para ver cuál es el candidato con más posibilidades de ganar. En otras palabras, estamos estudiando la variación que hay en preferencia de la población hábil para votar para poder sacar una conclusión (y el político poder estimar si necesita inventarse más mentiras para atraer a las presas).

Todo lo anterior aplica también para Machine Learning. Por ejemplo, los modelos de machine learning lo que buscan es predecir un resultado con cierto grado de confiabilidad. Pero, cómo puedo predecir un resultado si este puede variar en función de distintos factores? ¿Qué técnicas o herramientas puedo usar para poder buscar patrones en esa variación, que tal vez como humanos no podemos identificar, y así predecir un resultado?

¡Ajá! ¡La estadística!

Creo que ya entendemos bien el concepto, ahora pasemos a algunas técnicas estadísticas. Posiblemente algunas las conoces y otras no.

#### **Introducción a la Estadística Descriptiva**

La estadística descriptiva es la parte de la estadística que resume un conjunto de datos dado. Este conjunto de datos puede venir de una población o de una muestra. Pongamos un pequeño ejemplo:

Supongamos que tu jefe te dio un excel con los siguientes datos. Estos datos son los tiempos de respuesta de un servicio de atención al cliente y están expresados en minutos:

In [18]:

tiempos  = [15,10,80,10,30,25,12,10,9,12,45,10,48,20,32,33,10,52,100,10,22,23,14,27,5,8,9]

'''
¿Qué significa ese código?

En python esto se conoce como una lista y justamente lo que contiene es una lista de núnmeros (que en nuestro caso representan tiempos). Esa lista de números
es almacenada en una variable llamada "tiempos". Una lista en python puede estar compuesta de números, letras, otras listas, etc.
'''

None

Viendo los datos tal y cómo están presentados, ¿qué conclusión podrías sacar? Está difícil poder sacar una conslusión sin procesarlos de alguna manera, ¿verdad?

Aquí es donde entra la estadística descriptiva. La idea es poder resumir esos datos para que sean mucho más digerible y podamos sacar conclusiones a partir de ellos. Así que veamos algunas técnicas para poder resumir estos datos.

Estas técnicas se dividen en medidas de tendencia central y medidas de dispersión. Comencemos viendo las medidas de tendencia central.

#### **Medidas de Tendencia Central**

##### **La media**

La media o promedio es el estadístico más popular, y el más abusado (pobrecito, no le pagan suficiente snif snif). 

**WARNING - TURBULENCE ZONE**

La media lo que hace es buscar el punto donde se concentran la mayoría de los datos. Podemos ver la media como un centro de gravedad. Por ejemplo, la "media" de nuestro cuerpo viene siendo más o menos el área por donde está el ombligo, ya que esa es la zona donde se concentra todo nuestro peso.

Lo interesante de ver la media como el centro de gravedad de los datos es que no necesariamente la media será el centro. Volvamos al centro de gravedad.

Imáginate a un Gymholic que solo haga ejercicios de pecho y olvida el leg day. Llega a un punto que parece un globo inflado, un Johny Bravo hecho real, y todo su peso se concentra en esa zona (Tienes que imaginarte algo bien exagerado para que el ejemplo funcione. Algo como el pecho de Hulk, pero de la cintura para bajo es como una barbie). Eso significa que su centro de gravedad va a estar arriba del ombligo, porque ahí es donde está la mayor concentración de peso.

Con la media ocurre lo mismo. Esta se va a mover a donde esté la mayor cantidad de datos. Pongamos dos ejemplos numéricos. Vamos a suponer que estás haciendo un estudio de cuánto ganan los estudiantes de MNA y te pasan una lista de salarios (en USD) para que saques la media:

In [3]:
#Importamos la librería numpy para usar sus funciones. La librería se llama numpy pero para facilidad queremos apodarla como np. 
#La podemos apodar como queramos. Si quieres, le puedes llamar pepito y quedaría como import numpy as pepito
import numpy as np 

salarios = [100,150,130,115,100,113,145,132] #Definimos otra lista de python

np.mean(salarios) #Usamos el método mean (media) para que calcule la media de la lista definida arriba.



123.125

El resultado anterior nos da 123.125, lo cual sería el centro de gravedad de nuestros datos, y tiene sentido si miramos el conjunto de datos. Todos los vamores mas o menos se concentran alrededor de ese número. 

Posiblemente ahora te estés preguntando "¿Por qué estás complicando un concepto tan simple que uso todos los días?"

Porque la media tiene un serio problema, y es que es sensible a valores atípicos. Es justo en esta parte donde hace sentido el concepto de centro de gravedad. Volvamos al ejemplo anterior, pero ahora supongamos que Bill Gates decidió cursar la maestría con nosotros (y también aplicó para la beca de trayectoria profesional y le dieron un 30%).

Como tenemos un nuevo estudiante con nosotros, debemos actualizar nuestra lista anterior con el salario de Bill Gates:

In [4]:

#Como salarios es una lista de python, tengo acceso al método append. 
# Esto lo que hace es anexar al final de la lista el valor que está dentro del paréntesis
salarios.append(1500000000)

#Vamos a calcular la media con el nuevo valor
np.mean(salarios)

166666776.1111111

Mira el resultado de la media. ¿Crees que el salario promedio mensual de USD 166,666,776.11 representa la realidad de nosotros? Pues no! Somos unos pobretones alante de Bill Gates. Pero como vimos arriba, la media va a tender donde se concentre la mayor cantidad de información, y en este caso la mayor cantidad de información la contiene un solo dato de todo el conjunto: el salario de Billy.

Así que la próxima vez que uses la media, ten en cuenta la presencia de valores atípicos.

##### **La Mediana**

La mediana es otra medida de tendencia central y, a diferencia de la media, no es susceptible ante la presencia de valores atípicos. La mediana viene siendo el dato que está justamente en el centro, independendientemente de lo que haya en los extremos.

Tomemos el ejemplo anterior, incluyendo el salario de Billy y calculemos la mediana.


In [5]:
#Ahora usamos el método "median" sobre la lista "salarios" para extraer la mediana
np.median(salarios)

130.0

La mediana es USD 130. ¿Cuál crees que represente mejor el salario de la población? Los más de USD 160 MM que arrojó la media o la pírrica suma de USD 130 que nos dio la mediana?.

La razón por que la mediana funciona mejor ante presencia de atípicos, es porque para calcularse debes organizar los datos de menor a mayor (o viceversa) y luego buscar el valor que está justamente en el medio. Por tanto, no importa que tantos valores atípicos tengas a la derecha o a la izquierda del conjunto, al final lo que buscas es el valor del centro.

**TIP**

Cuando tengas un conjunto de datos (incluyendo el que te pase tu jefe), calcula tanto la media y la mediana. Si ambas tienen valores similares, significa que tu conjunto de datos no tiene valores atípicos y no varía mucho. Si la media y mediana son muy distantes (como el caso anterior), significa que tu conjunto de datos tiene atípicos o mucha variación (es decir, es poco consistente).

##### **La Moda**

La moda es el valor que más se repite. El inconveniente con la moda es que puede que tu conjunto de datos no tenga moda o tenga más de una moda.

Cuando tienes más de una moda se dice que tienes una distribución bimodal (si son dos) o multimodal (si son más de dos). Cuando graficas estos datos con un histograma, vas a notar que hay dos picos en el caso bimodal o múltiples picos.

Cuando tus datos tienen más de una moda significa que tienes grupos mezclados en un mismo conjunto de datos. Por ejemplo, volvamos al caso de los salarios y extendamos la lista con nuevos valores.

In [6]:
import statistics #Estamos importando otro módulo porque numpy no tiene una función que necesito para ilustrar el concepto

#Aquí usamos el método extend de la lista. 
#Esto lo que hace es recibir como entrada una lista que quieras anexar al final de salarios. 
#En otras palabras, los valores que ves se añadieron a la lista que ya habíamos definido, preservando los valores anteriores.
salarios.extend([130,130,130,1000,1000,1000,1000])

#Usamos el método multimode para calcular los valores que más se repiten.
statistics.multimode(salarios)

[130, 1000]

Fíjate que con los nuevos datos que ingresamos ahora existen dos modas, una de 130 y otra de 1000. Una posible explicación es que hay dos grupos de salarios: Los pobretones de USD 130 y los ricachones que reciben USD 1000. En algunos escenarios, tal vez sea mejor separar los grupos para analizarlos por separados. En otros casos tal vez sea necesario estratificar para generar muestras homogéneas, en otras tal vez se tengan que transformar los datos, etc. Todo dependerá de lo que quieras hacer.

#### **Medidas de Dispersión**

Las medidas de dispersión lo que hacen es determinar que tanta variación hay en nuestro conjunto de datos. En otras palabras, las medidas de dispersión nos dan una idea de que tan consistente son los datos de nuestro conjunto. 

**La Desviación Estándar (DE)**

Para ponerlo en términos simples, la DE se podría decir que:

**TURBULENCE AHEAD** 

**Es la distancia promedio de cada valor con respecto a su media.**

Vamos a ver con qué se come eso.

Tenemos el siguiente conjunto de datos:

[1, 2, 3, 4, 5]

Si calculas la media verás que te da un valor de 3 (no me creas, busca una calculadora y confírmalo. No seas haragán).

Pero supongamos que ahora queremos saber que tan dispersos están los valores entre uno y otro, pero para poder hacer esto necesitamos tener una referencia contra la cual medir esta distancia. Así que tomamos la media como punto de referencia.

**NOTA:** la DE no se calcula exactamente como lo muestro a continuación. Simplemente hago el ejercicio para entender el concepto. Más adelante veremos la fórmula verdadera.

Calculemos la distancia de manera manual (cáculalo tu también. Que no te de flojera):

* La distancia entre 1 y 3 es 2
* La distancia entre 2 y 3 es 1
* La distancia entre 3 y 3 es 0
* La distencia entre 4 y 3 es 1
* La distancia entre 5 y 3 es 2

Ahora, vamos a promediar esas distancias: (2 + 1 + 0 + 1 +2)/5 = 1.2

Ese 1.2 vendría siendo algo parecido a la DE estándar. Significa que los datos tienen en promedio una distancia de 1.2 unidades con respecto a su media. Mientras más grande es este número, mayor dispersión. Mientras más pequeño, menor dispersión (y por tanto mayor consistencia).

Veamos cuál es la DE de nuestro conjunto de salarios.

In [8]:
#Como hemos pullado tanto el conjunto salarios, preferí redefinirlo a sus valores originales.
salarios = [100,150,130,115,100,113,145,132] 

np.std(salarios)

17.90556826799976

La DE en este caso es de 17.90. Eso significa que los salarios varían en promedio USD 17.90 con respecto a la media (recuerda, el punto de referencia es la media).

Ahora veamos que pasa si incluimos el salario de nuestro amigo Billy:

In [10]:
salarios.append(1500000000)

np.std(salarios)

471404482.0965775

Uuuf... se disparó. Eso significa que la DE tiene el mismo problema de la media, se verá afectada ante valores atípicos.

Tiene sentido puesto que la DE usa como valor de referencia la media. Por tanto, si la media está podrida, pues la DE también lo estará.

##### **Los Percentiles**

Los percentiles son otra forma de analizar los datos y consiste en dividir el conjunto de datos en 100 partes iguales (por eso el nombre de percentil). 

Por ejemplo, supongamos que queremos calcular el percentil 75. Esto significa que queremos encontrar el valor umbral donde se encuentra el 75% de los datos.

Veamos un ejemplo con nuestro conjunto de salarios y busquemos justamente el percentil 75.

In [14]:
np.percentile(salarios, q = 75)

145.0

El percentil 75 es el valor de USD 145.00. Esto significa que el 75% de los datos caen por debajo del salario de los USD 145.00 (esperemos que este número se infle cuando terminemos MNA).

Probemos otros valores a ver que pasa:

In [16]:
'''
Aquí estamos usando la función print para imprimir datos en la consola. Te podrías preguntar
que qué significa esa f delante del texto. Esa f significa format y es para poder mezclar texto con variables y aplicar formatos.

Siempre que la f esté delante, puedes poner entre {} variables o cálculos simples para imprimir en la consola.

Dentro de los {} se encuentra el método percentile de la librería numpy. Le paso a este método el conjunto de datos "salarios" y el percentil que quiero
calcular como parte del parámetro "q" (ni idea por qué le llamaron así. Tal vez de quantile)
'''

print(f"Percentil 25: {np.percentile(salarios, q = 25)}")
print(f"Percentil 50: {np.percentile(salarios, q = 50)}")
print(f"Percentil 75: {np.percentile(salarios, q = 75)}")

Percentil 25: 113.0
Percentil 50: 130.0
Percentil 75: 145.0


Lo que hicimos arriba fue calcular lo que se llaman los cuartiles, los cuales son el percentil 25 (primer cuartil), el percentil 50 (segundo cuartil o mediana) y el percentil 75 (tercer cuartil).

Fíjate que el percentil 50 es exactamente la mediana, y es que la mediana realmente es un tipo de percentil. 

La interpretación de los cuartiles se hace de la siguiente manera:

* El primer cuartil lo que te dice es que el 25% de los salarios están por debajo de USD 113.
* La mediana, que ya la vimos, dice que el 50% de los datos está por debajo de USD 130
* Mientras que el tercer cuartil te dice que el 75% de los datos está por debajo de USD 145.

¿No me crees que la mediana y el percentil 50 es lo mismo?

Vamos a calcular la mediana nuevamente

In [17]:
np.median(salarios)

130.0

Nada más que agregar Sr. Juez

##### **La Varianza**

¿Recuerdas el ejemplo con el conjunto [1, 2, 3, 4, 5], donde calculamos las distancias?

Bueno, pues está mal. Hay un pequeño error que cometimos y aquí te lo presento:


* La distancia entre 1 y 3 es -2
* La distancia entre 2 y 3 es -1
* La distancia entre 3 y 3 es 0
* La distencia entre 4 y 3 es 1
* La distancia entre 5 y 3 es 2

Fíjate que la distancia entre 1 y 3 es -2 porque 1 - 3 = -2.

Ahora tratemos de calcular la distancia promedio:

-2 -1 + 0 + 1 + 2 = 0

Que lío. Ahora todo se hace cero. 

Pero los matemáticos inventaron una forma de salirse debajo del camión: La Varianza.

Básicamente lo que esto hace es elevar todo al cuadrado para evitar que se cancelen los números (ya que todo número elevado al cuadrado es positivo):

$-2^{2}$ + $-1^{2}$ +  ... + $2^{2}$ = 10

Si calculamos la media, tendremos: $\frac{10}{5}$	 = 2

Pero tenemos un problema. El resultado que acabamos de calcular está al cuadrado, por lo que tenemos que hacer el proceso inverso calculando la raiz caudrada:

DE = $\sqrt{2}$ = 1.4

En pocas palabras, la varianza y la DE están muy relacionadas, puesto que una es el cuadrado de la otra.

La técnica de la varianza juega un papel muy importante en temas avanzados de estadística. Su fórmula podría variar dependiendo de la técnica, pero el concepto sigue siendo el mismo.

El único problema de la varianza es que su resultado no está en las unidades originales del conjunto de datos, por lo que dificulta su interpretación. En este caso, el resultado de la varianza fue 2, lo que significa USD 2 dólares-cuadrado.


##### **Rango**

El rango es una medida de dispersión muy sencilla de calcular. Básicamente consiste en restar el valor máximo del mínimo.

El rango, como dice su nombre, te indica que tan amplio es tu conjunto de datos. El gran problema con el rango es que es demasiado sensible a valores extremos (atípicos). De hecho, es más llorón que la DE.

Veamos un ejemplo:

In [21]:
np.max(salarios) - np.min(salarios)

1499999900

Como el conjunto de datos incluye el salario de nuestro amigo Billy, da un valor demasiado alto y que no representa la variabilidad promedio de nuestros datos. 

A continuación, veamos otra medida que nos podría ayudar ante esta situación.

##### **Rango Intercuartil**

Esta es otra medida de dispersión y básicamente consiste en calcular el rango que hay entre el cuartil 1 (percentil 25) y cuartil 3 (percentil 75).

Veamos un ejemplo:

In [20]:
IQR = np.percentile(salarios, q = 75) - np.percentile(salarios, q = 25)
IQR

32.0

Lo anterior significa que el rango de los datos que están en el centro es de USD 32.00. El rango intercuartil es el análogo a la mediana, y al igual que la mediana no es sensible a valores atípicos.

Fíjate que hicimos el cálculo con el salario de Billy incluido, y a pesar de eso nos dio un valor más razonable que el rango.

Por tanto, si tienes valores atípicos o mucha variabilidad en tus datos, tal vez sea mejor analizar el rango intercuartil.

#### **Población vs Muestra**


En la mayoría de los casos no podremos contar con todos los datos de la población y por tanto debemos estudiar una muestra, y con base en la muestra entonces inferenciar un parámetro de la población. Esto significa que además de la estadística descriptiva, también tenemos la estadística inferencial.

Ahora bien, los conceptos que vimos anteriormente aplican tanto para la estadística descriptiva como inferencial, sin embargo, algunas de las fórmulas podrían variar (como la DE) dependiendo si estamos trabajando con una muestra o una población. Es por ello que al referirnos a una media de la muestra la denotamos con un símbolo diferente a la media de la población.

A continuación te presento los símbolos que normalmente se usan:

* Media de la muestra: $\bar{X}$
* Media de la población $\mu$ (se pronuncia miu)
* Desviación Estándar de la muestra: s
* Desviación Estándar de la población: $\sigma$ (se pronuncia sigma)
* Varianza de la muestra: $s^{2}$
* Varianza de la población: $\sigma^{2}$
